glean-core/ffi/src/fd_logger.rs

```diff
@@ -0,0 +1,83 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::fs::File;
+use std::io::Write;
+use std::sync::RwLock;
+
+#[cfg(target_os = "windows")]
+use std::os::windows::io::FromRawHandle;
+
+#[cfg(target_os = "windows")]
+use std::ffi::c_void;
+
+#[cfg(not(target_os = "windows"))]
+use std::os::unix::io::FromRawFd;
+
+use serde::Serialize;
+
+/// An implementation of log::Log that writes log messages in JSON format to a
+/// file descriptor/handle. The logging level is ignored in this implementation:
+/// it is up to the receiver of these log messages (on the language binding
+/// side) to filter the log messages based on their level.
+/// The JSON payload of each message in an object with the following keys:
+///    - `level` (string): One of the logging levels defined here:
+///      https://docs.rs/log/0.4.11/log/enum.Level.html
+///    - `message` (string): The logging message.
+pub struct FdLogger {
+    pub file: RwLock<File>,
+}
+
+#[derive(Serialize)]
+struct FdLoggingRecord {
+    level: String,
+    message: String,
+}
+
+#[cfg(target_os = "windows")]
+fn get_file_from_fd(fd: u64) -> File {
+    unsafe { File::from_raw_handle(fd as *mut c_void) }
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_file_from_fd(fd: u64) -> File {
+    unsafe { File::from_raw_fd(fd as i32) }
+}
+
+impl FdLogger {
+    pub fn new(fd: u64) -> Self {
```

Comment:
And then this also becomes `pub unsafe fn new`.

And yes, that means this unsafe propagates upwards, but there's not much we can do.
From the docs:

> This function is also unsafe as the primitives currently returned have the contract that they are the sole owner of the file descriptor they are wrapping. Usage of this function could accidentally allow violating this contract which can cause memory unsafety in code that relies on it being true.

and there is no way for us to check this precondition, so we have to mark it unsafe.
The good thing is: this is already called from an `extern "C"` function, and C land is unsafe, so it's unsafe all the way ~~down~~ up!